### ISSUE TITLE
API for getting hidden layer state

### ISSUE THREAD

**Issue by goncalopp** [tensorflow/tensorflow#3658]
I'm using `tf.contrib.learn.TensorFlowDNNRegressor` and I'd like to get the hidden layers state after forward propagation (calling predict()). There doesn't seem to be an API to do this. Will one be provided in the future, or would I need to avoid `tf.contrib.learn`?

**Comment by ilblackdragon** [comment]
@goncalopp We don't have anything planned at this point. One way around is to use custom `Estimator` class and return various tensors that you want to inspect in the `predictions` dictionary.

Would this work for you?

PS. you should switch to use `DNNRegressor` as `TensorFlowDNNRegressor` is deprecated.

**Comment by goncalopp** [comment]
Thanks for looking into this!
I've been looking into this, and writing a custom Estimator would involve essentially copy-pasting 
`_infer_model()`, `_get_predict_ops()` and `_logits()`, which would be very brittle, considering how fast this code is changing. Would you be interested in a pull request that modifies `BaseEstimator` to optionally return these?

**Comment by ilblackdragon** [comment]
The better solution is to add hooks into predict method, to allow to pass `CaptureVariable` that would capture the values of hidden layers. 
+ @ispirmustafa who is modifying underlaying layers to use `MonitoredSession` and can easily add this.

**Comment by drpngx** [comment]
We changed things around with the MonitoredSession if memory serves. @goncalopp is this still current?

**Comment by goncalopp** [comment]
@drpngx I've moved to using a custom `Estimator`, as ` ilblackdragon suggested - it's not as easy, but works well. We can close this, if it's not a priority.
I took a quick look on `MonitoredSession`, but it seems geared to persisting state, not accessing it? The feature request here is accessing the raw neuron weights, post-training - for visualization purposes, for example. Would `MonitoredSession` be able to do that?
